xc/device.ml

```diff
@@ -1409,29 +1409,46 @@ module Vkbd = struct
 end
 
 module Vusb = struct
-  let call_usb_reset hostbus hostport =
-    let usb_reset_script = "/opt/xensource/libexec/usb_reset.py" in
+  let exec_usb_reset_script argv =
+    let stdout, stderr = Forkhelpers.execute_command_get_output !Xc_resources.usb_reset_script argv in
+    debug "usb_reset_script %s returned stdout=%s stderr=%s" (String.concat " " argv) stdout stderr
+
+  let cleanup domid =
+    try
+      let cmd = ["cleanup"; "-d"; string_of_int domid] in
+      exec_usb_reset_script cmd
+    with _ ->
+      warn "Failed to clean up VM %s" (string_of_int domid)
+
+  let call_usb_reset hostbus hostport domid pid action priv =
     try
-      info "Reset USB device with hostbus %s and hostport %s before passthrough to vm." hostbus hostport;
-      ignore (Forkhelpers.execute_command_get_output usb_reset_script [ hostbus ^ "-" ^ hostport ])
+      match action, priv with
+      | "attach", false ->
+        let cmd = [ "attach"; hostbus ^ "-" ^ hostport; "-d"; string_of_int domid; "-p"; string_of_int pid] in
+        exec_usb_reset_script cmd
+      | "attach", true ->
+        (*When qemu is privileged, attach will only reset device.*)
+        let cmd = [ "attach"; hostbus ^ "-" ^ hostport; "-d"; string_of_int domid; "-p"; string_of_int pid; "-r"] in
+        exec_usb_reset_script cmd
+      | "detach", false ->
+        let cmd = [ "detach"; hostbus ^ "-" ^ hostport ; "-d"; string_of_int domid; "-i"; "0" ; "0"; ] in
+        exec_usb_reset_script cmd
+      | "detach", true -> () (*When qemu is privileged, detach will do nothing.*)
+      | _, _ -> () (*Invalid operation, will do nothing*)
     with _ ->
-      error "Failed to reset USB device with hostbus %s and hostport %s." hostbus hostport;
+      error "Failed to %s USB device with hostbus %s and hostport %s." action hostbus hostport;
       failwith "Call to usb reset failed."
```

Comment:
Why `failwith` and not something more tractable like `raise (Internal_error ...`?